source/common/stats/thread_local_store.cc

```diff
@@ -204,29 +253,123 @@ Histogram& ThreadLocalStoreImpl::ScopeImpl::histogram(const std::string& name) {
+  std::string tag_extracted_name = parent_.getTagsForName(final_name, tags);
+  if (!central_ref) {
+    // Since MetricImpl only has move constructor, we are explicitly copying here.
+    std::string central_tag_extracted_name(tag_extracted_name);
+    std::vector<Tag> central_tags(tags);
+    central_ref.reset(new ParentHistogramImpl(final_name, parent_, *this,
+                                              std::move(central_tag_extracted_name),
+                                              std::move(central_tags)));
+  }
+  return *central_ref;
+}
+
+Histogram& ThreadLocalStoreImpl::ScopeImpl::tlsHistogram(const std::string& name) {
+  // Here prefix will not be considered because, by the time ParentHistogram calls this method
+  // during recordValue, the prefix is already attached to the name.
+  TlsHistogramSharedPtr* tls_ref = nullptr;
   if (!parent_.shutting_down_ && parent_.tls_) {
-    tls_ref = &parent_.tls_->getTyped<TlsCache>().scope_cache_[this].histograms_[final_name];
+    tls_ref = &parent_.tls_->getTyped<TlsCache>().scope_cache_[this].histograms_[name];
   }
 
   if (tls_ref && *tls_ref) {
     return **tls_ref;
   }
 
   std::unique_lock<std::mutex> lock(parent_.lock_);
-  HistogramSharedPtr& central_ref = central_cache_.histograms_[final_name];
-  if (!central_ref) {
-    std::vector<Tag> tags;
-    std::string tag_extracted_name = parent_.getTagsForName(final_name, tags);
-    central_ref.reset(
-        new HistogramImpl(final_name, parent_, std::move(tag_extracted_name), std::move(tags)));
-  }
+  std::vector<Tag> tags;
+  std::string tag_extracted_name = parent_.getTagsForName(name, tags);
+  TlsHistogramSharedPtr hist_tls_ptr(
+      new ThreadLocalHistogramImpl(name, std::move(tag_extracted_name), std::move(tags)));
+  ParentHistogramImplSharedPtr& central_ref = central_cache_.histograms_[name];
+  central_ref->addTlsHistogram(hist_tls_ptr);
 
   if (tls_ref) {
-    *tls_ref = central_ref;
+    *tls_ref = hist_tls_ptr;
   }
+  return *hist_tls_ptr;
+}
 
-  return *central_ref;
+ThreadLocalHistogramImpl::ThreadLocalHistogramImpl(const std::string& name,
+                                                   std::string&& tag_extracted_name,
+                                                   std::vector<Tag>&& tags)
+    : MetricImpl(name, std::move(tag_extracted_name), std::move(tags)), current_active_(0),
+      flags_(0), created_thread_id_(std::this_thread::get_id()) {
+  histograms_[0] = hist_alloc();
+  histograms_[1] = hist_alloc();
+}
+
+ThreadLocalHistogramImpl::~ThreadLocalHistogramImpl() {
+  hist_free(histograms_[0]);
+  hist_free(histograms_[1]);
+}
+
+void ThreadLocalHistogramImpl::recordValue(uint64_t value) {
+  ASSERT(std::this_thread::get_id() == created_thread_id_);
+  hist_insert_intscale(histograms_[current_active_], value, 0, 1);
+  flags_ |= Flags::Used;
+}
+
+void ThreadLocalHistogramImpl::merge(histogram_t* target) {
+  const uint64_t other_histogram_index = otherHistogramIndex();
```

Comment:
optional microoptimization: capture the histogram in a temp rather than the index:
```
histogram_t** other_histogram = &histograms_[otherHistogramIndex()];
hist_accumulate(target, other_histogram, 1);
hist_clear(target, *other_histogram);
```